src/cpu/simple/probes/looppoint_analysis.cc

```diff
@@ -0,0 +1,280 @@
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "cpu/simple/probes/looppoint_analysis.hh"
+
+namespace gem5
+{
+
+LooppointAnalysis::LooppointAnalysis(const LooppointAnalysisParams &params)
+    : ProbeListenerObject(params),
+      lpaManager(params.looppoint_analysis_manager),
+      bbValidAddrRange(params.bb_valid_addr_range),
+      markerValidAddrRange(params.marker_valid_addr_range),
+      ifListening(params.if_listening),
+      bbInstCounter(0)
+{
+    DPRINTF(LooppointAnalysis, "Start listening from the beginning of the "
+                            "simulation? %s\n", ifListening ? "Yes" : "No");
+
+    for (int i = 0; i < params.bb_excluded_addr_ranges.size(); i++)
+    {
+        bbExcludedAddrRanges.push_back(
+            AddrRange(
+                params.bb_excluded_addr_ranges[i].start(),
+                params.bb_excluded_addr_ranges[i].end()
+            )
+        );
+        DPRINTF(LooppointAnalysis, "Excluding address range: (%li, %li)\n",
+                params.bb_excluded_addr_ranges[i].start(),
+                params.bb_excluded_addr_ranges[i].end()
+        );
+    }
+
+    DPRINTF(LooppointAnalysis, "Valid address range: (%li, %li)\n",
+            bbValidAddrRange.start(), bbValidAddrRange.end());
+    DPRINTF(LooppointAnalysis, "Valid marker address range: (%li, %li)\n",
+            markerValidAddrRange.start(), markerValidAddrRange.end());
+}
+
+void
+LooppointAnalysis::updateLocalBBV(const Addr pc)
+{
+    if (localBBV.find(pc) == localBBV.end())
+    {
+        localBBV.insert(std::make_pair(pc, 1));
+    }
+    else
+    {
+        localBBV.find(pc)->second++;
+    }
```

Comment:
I'm not going to do this for the whole file, but please change this style

```suggestion
    if (localBBV.find(pc) == localBBV.end()) {
        localBBV.insert(std::make_pair(pc, 1));
    } else {
        localBBV.find(pc)->second++;
    }
```